### DISCUSSION TITLE
Image Registry configuration patch isn't applied when adding `allowedRegistries`

### DISCUSSION THREAD

**Discussion by ProtossGP32** [okd-project/okd#1646] [General]
**Describe the bug**

I'd llke to add the internal registry as an additional registry where to pull images from when launching Builds (for example, to not depend on docker pull quotas). For that, I've followed the steps in [the official docs](https://docs.okd.io/latest/openshift_images/image-configuration.html#images-configuration-file_image-configuration), but after finished patching the `image.config.openshift.io/cluster` custom resource, the nodes aren't cordoned and the changes only remain in there, there's no rollout of it.

**Version**

OKD version release: 4.13.0-0.okd-2023-05-22-052007
Install method: IPI
Cloud platform: OpenStack

**How reproducible**

100%, just follow the instructions in [the official docs](https://docs.okd.io/latest/openshift_images/image-configuration.html#images-configuration-file_image-configuration) to add the internal registry as an `allowedRegistries` entry:
```
oc edit image.config.openshift.io/cluster
```
And then adding the list of allowed registers:
```yaml
 apiVersion: config.openshift.io/v1
 kind: Image
 [...]
 spec:
   registrySources:
     allowedRegistries:
     - image-registry.openshift-image-registry.svc:5000
     - quay.io
     - registry.redhat.io
     - docker.io
```

**Expected behaviour**
Nodes are cordoned, new config is applied and nodes are uncordoned again. I'd expect to see the `NotReady` status in the nodes for some amount of time.

**Real behaviour**
Nodes aren't cordoned at all, they keep on being `Ready`; the custom resource is modified (checked again with `oc get image.config.openshift.io/cluster -o yaml`) and thus there's no rollout of the changes in the platform.

**Log bundle**

```
ClusterID: ea345ff1-460d-4ece-8884-540d0fc58702
ClusterVersion: Stable at "4.13.0-0.okd-2023-05-22-052007"
ClusterOperators:
        clusteroperator/openshift-samples is degraded because Samples installation in error at 4.13.0-0.okd-2023-05-22-052007: &errors.errorString{s:"global openshift image configuration prevents the creation of imagestreams using the registry "}
```
- `.openshift_install.log`: https://gist.github.com/ProtossGP32/0efe2200acfbffb3e13601a9503d561f
- Download `must-gather` archive: [LINK](https://we.tl/t-HtBtU6eBsD)

**Comment by ProtossGP32** [comment]
@vrutkovs I'd like to know why this issue has been moved to a discussion, as it's the second time that happens to me. What would be the exact criteria for bugs? Right now I followed an official procedure and didn't return the expected results, so I think of this as a bug.

**Reply by vrutkovs** [reply]
> this issue has been moved to a discussion

No log bundle / must gather provided, for starters. Also, issues are for actual confirmed bugs, so once we confirm that this is indeed an OKD-specific issue we need to fix (and not a documentation / OCP bug) we'll move it to issues again

**Reply by ProtossGP32** [reply]
Thanks for the pointers! I've updated the description with both the `.openshift_installer.log` and the `must-gather` output (WeTransfer link, expires in 7 days).

**Comment by ProtossGP32** [comment]
It's been two weeks since I opened this discussion, and I'm struggling with the same issue. Image registry config isn't updated in the nodes and I can't access the internal registry to pull images from there. Does anybody have any advice on how to further proceed with the troubleshooting?
Thanks!

**Reply by dwkop** [reply]
Did you check if a new MachineConfig was rendered?
This actually kicking in the process of rolling out new changes to the nodes.
So "Compute->MachineConfigs>RenderedMachineConfig sort desc".

Also check MachineConfigPool if they are in an "update" state.

**Reply by ProtossGP32** [reply]
Hi @dan1el-k , thanks for the reply! I've tried once again to edit the config and check if it kicks the rollouts you mentioned:
- Indeed, when modifying the `image.config.openshift.io/cluster` config, two new MachineConfigs are created (edit done on 2023/07/20 15:08):
![imatge](https://github.com/okd-project/okd/assets/19353116/faaf6fe3-888a-41f1-a6a5-2cc14d4a59aa)

- The MachineConfigPools for master and worker show changes, and it seems that now they should point towards the new MachineConfigs:
![imatge](https://github.com/okd-project/okd/assets/19353116/974a428f-838c-455e-8aea-f2e2fc0e6dac)
![imatge](https://github.com/okd-project/okd/assets/19353116/fd21ca2d-333d-4007-aa46-fff637f0fa56)

- After a second update of the image.config.openshift.io/cluster, now I finally see the Updating status on both MachineConfigPools:
![imatge](https://github.com/okd-project/okd/assets/19353116/72484bea-6d31-42de-a464-075f09a9244b)

Do you know how I can check if the changes have been applied without having to spawn a new build?

**Reply by ProtossGP32** [reply]
I've tried creating a pod that should have to pull the image from the internal registry with no success. It looks for the docker image name in the default registries:
- registry.access.redhat.com
- docker.io
But not on the additional ones when I modified the `image.config.openshift.io/cluster`:
    - quay.io
    - registry.redhat.io
    - image-registry.openshift-image-registry.svc:5000
    - docker.io

```
Jul 20 13:34:32.608582 okd-cluster-nwhj5-worker-0-75jlr.novalocal crio[3103]: time="2023-07-20 13:34:32.608415416Z" level=info msg="Checking image status: custom-project/custom-docker-image:latest" id=993d7259-cc2f-46f2-a8d3-f3b30fd2c2da name=/runtime.v1.ImageService/ImageStatus
Jul 20 13:34:32.609249 okd-cluster-nwhj5-worker-0-75jlr.novalocal crio[3103]: time="2023-07-20 13:34:32.608766724Z" level=info msg="Resolving "custom-project/custom-docker-image\" using unqualified-search registries (/etc/containers/registries.conf)"
Jul 20 13:34:32.609249 okd-cluster-nwhj5-worker-0-75jlr.novalocal crio[3103]: time="2023-07-20 13:34:32.609041659Z" level=info msg="Image custom-project/custom-docker-image:latest not found" id=993d7259-cc2f-46f2-a8d3-f3b30fd2c2da name=/runtime.v1.ImageService/ImageStatus
Jul 20 13:34:32.610206 okd-cluster-nwhj5-worker-0-75jlr.novalocal crio[3103]: time="2023-07-20 13:34:32.610033001Z" level=info msg="Pulling image: custom-project/custom-docker-image:latest" id=534d9ea8-c5dd-4bf5-a74d-87b2ee8d2321 name=/runtime.v1.ImageService/PullImage
Jul 20 13:34:32.610341 okd-cluster-nwhj5-worker-0-75jlr.novalocal crio[3103]: time="2023-07-20 13:34:32.610221592Z" level=info msg="Resolving \"custom-project/custom-docker-image\" using unqualified-search registries (/etc/containers/registries.conf)"
Jul 20 13:34:32.635282 okd-cluster-nwhj5-worker-0-75jlr.novalocal crio[3103]: time="2023-07-20 13:34:32.635211417Z" level=info msg="Trying to access \"registry.access.redhat.com/custom-project/custom-docker-image:latest\""
Jul 20 13:34:33.078432 okd-cluster-nwhj5-worker-0-75jlr.novalocal crio[3103]: time="2023-07-20 13:34:33.078357981Z" level=info msg="Trying to access \"docker.io/custom-project/custom-docker-image:latest\""
Jul 20 13:34:33.608618 okd-cluster-nwhj5-worker-0-75jlr.novalocal kubenswrapper[3133]: E0720 13:34:33.608539 3133 dns.go:156] "Nameserver limits exceeded" err="Nameserver limits were exceeded, some nameservers have been omitted, the applied nameserver line is: 10.0.27.138 10.0.27.4 84.88.52.35"
Jul 20 13:34:34.457891 okd-cluster-nwhj5-worker-0-75jlr.novalocal kubenswrapper[3133]: E0720 13:34:34.457801 3133 remote_image.go:171] "PullImage from image service failed" err="rpc error: code = Unknown desc = reading manifest latest in docker.io/custom-project/custom-docker-image: requested access to the resource is denied" image="custom-project/custom-docker-image:latest"
Jul 20 13:34:34.457891 okd-cluster-nwhj5-worker-0-75jlr.novalocal kubenswrapper[3133]: E0720 13:34:34.457895 3133 kuberuntime_image.go:53] "Failed to pull image" err="rpc error: code = Unknown desc = reading manifest latest in docker.io/custom-project/custom-docker-image: requested access to the resource is denied" image="custom-project/custom-docker-image:latest"
Jul 20 13:34:34.458276 okd-cluster-nwhj5-worker-0-75jlr.novalocal kubenswrapper[3133]: E0720 13:34:34.458102 3133 kuberuntime_manager.go:872] container &Container{Name:custom-docker-image,Image:custom-project/custom-docker-image:latest,Command:[],Args:[],WorkingDir:,Ports:[]ContainerPort{ContainerPort{Name:,HostPort:0,ContainerPort:8080,Protocol:TCP,HostIP:,},},Env:[]EnvVar{},Resources:ResourceRequirements{Limits:ResourceList{},Requests:ResourceList{},Claims:[]ResourceClaim{},},VolumeMounts:[]VolumeMount{VolumeMount{Name:kube-api-access-5kbgz,ReadOnly:true,MountPath:/var/run/secrets/kubernetes.io/serviceaccount,SubPath:,MountPropagation:nil,SubPathExpr:,},},LivenessProbe:nil,ReadinessProbe:nil,Lifecycle:nil,TerminationMessagePath:/dev/termination-log,ImagePullPolicy:Always,SecurityContext:&SecurityContext{Capabilities:&Capabilities{Add:[],Drop:[ALL],},Privileged:nil,SELinuxOptions:nil,RunAsUser:*1000620000,RunAsNonRoot:nil,ReadOnlyRootFilesystem:nil,AllowPrivilegeEscalation:*false,RunAsGroup:nil,ProcMount:nil,WindowsOptions:nil,SeccompProfile:nil,},Stdin:false,StdinOnce:false,TTY:false,EnvFrom:[]EnvFromSource{},TerminationMessagePolicy:File,VolumeDevices:[]VolumeDevice{},StartupProbe:nil,} start failed in pod example_custom-project(05109abf-aa8e-4660-b6b8-814fe1430ff8): ErrImagePull: rpc error: code = Unknown desc = reading manifest latest in docker.io/custom-project/custom-docker-image: requested access to the resource is denied
Jul 20 13:34:34.458276 okd-cluster-nwhj5-worker-0-75jlr.novalocal kubenswrapper[3133]: E0720 13:34:34.458172 3133 pod_workers.go:965] "Error syncing pod, skipping" err="failed to \"StartContainer\" for \"custom-docker-image\" with ErrImagePull: \"rpc error: code = Unknown desc = reading manifest latest in docker.io/custom-project/custom-docker-image: requested access to the resource is denied\"" pod="incisive/example" podUID=05109abf-aa8e-4660-b6b8-814fe1430ff8
Jul 20 13:34:34.916469 okd-cluster-nwhj5-worker-0-75jlr.novalocal ovs-vswitchd[1153]: ovs|77416|connmgr|INFO|br-ex<->unix#488069: 2 flow_mods in the last 0 s (2 adds)
```

**Reply by ProtossGP32** [reply]
Hi again,
I've also ssh-ed inside the nodes to see the current status of the `/etc/containers/registries.conf` file and they appear untouched, not including the additional unqualified registries in the edit:
```
[core@okd-cluster-nwhj5-master-0 ~]$ cat /etc/containers/registries.conf | grep -v ^#
unqualified-search-registries = ['registry.access.redhat.com', 'docker.io']
```

**Reply by ProtossGP32** [reply]
Further investigation on the open enhancements in OpenShift has led me to find these user stories that are exactly what I'd expect from the platform in terms of image registry access:
[Add unqualified search registries - User Stories](https://github.com/openshift/enhancements/blob/master/enhancements/api-review/add-unqualified-search-registries.md#user-stories)

What should I have to do in order to achieve a similar outcome, knowing that this doesn't seem to be fully implemented yet?

**Reply by vrutkovs** [reply]
see https://access.redhat.com/solutions/5063741:
```
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfig
metadata:
  labels:
    machineconfiguration.openshift.io/role: worker
  name: 99-worker-unqualified-search-registries
spec:
  config:
    ignition:
      version: 2.2.0
    storage:
      files:
      - contents:
          source: data:text/plain;charset=utf-8;base64,dW5xdWFsaWZpZWQtc2VhcmNoLXJlZ2lzdHJpZXMgPSBbJ3F1YXkuaW8nXQo=
        filesystem: root
        mode: 0420
        path: /etc/containers/registries.conf.d/99-worker-unqualified-search-registries.conf
```

**Reply by ProtossGP32** [reply]
I'm sorry, I can't view the full Red Hat solution as I don't have a subscription. What does this MachineConfig tries to do that the changes in the image registry config doesn't do? Do I have to provide the base64 conversion of the expected registries.conf with my own registries here?

**Reply by LorbusChris** [reply]
The RH solution links to https://docs.openshift.com/container-platform/4.13/openshift_images/image-configuration.html

see https://docs.okd.io/4.13/openshift_images/image-configuration.html for the OKD variant of the docs

**Reply by ProtossGP32** [reply]
Hi @LorbusChris , thanks for the quick reply! The docs linked in the solution are in fact the ones I followed for the whole registry configuration (in fact I open this discussion pointing towards them). In these links there's no mention to the MachineConfig resource that @vrutkovs shared on his last message, so how are they related to each other? I understand that this `MachineConfig` resource should manually add the path to the unqualified search registries that should have been already created when modifying the `image.config.openshift.io/cluster`? Sorry about being so vague at this point of the discussion, but I feel like walking in circles since a long time just to allow the system to access its own internal registry
